perf/uncore_libpfm.go

```diff
@@ -246,147 +322,187 @@ func (c *uncoreCollector) Destroy() {
-				klog.V(5).Infof("Closing uncore perf_event file descriptor for event %q, PMU %s and CPU %d", name, pmu, cpu)
-				err := file.Close()
-				if err != nil {
-					klog.Warningf("Unable to close perf_event file descriptor for event %q, PMU %s and CPU %d", name, pmu, cpu)
+	for groupIndex, groupPMUs := range c.cpuFiles {
+		for pmu, group := range groupPMUs {
+			for name, cpus := range group.cpuFiles {
+				for cpu, file := range cpus {
+					klog.V(5).Infof("Closing uncore perf_event file descriptor for event %q, PMU %s and CPU %d", name, pmu, cpu)
+					err := file.Close()
+					if err != nil {
+						klog.Warningf("Unable to close perf_event file descriptor for event %q, PMU %s and CPU %d", name, pmu, cpu)
+					}
 				}
+				delete(group.cpuFiles, name)
 			}
-			delete(pmus, pmu)
+			delete(groupPMUs, pmu)
 		}
-		delete(c.cpuFiles, name)
+		delete(c.cpuFiles, groupIndex)
 	}
 }
 
 func (c *uncoreCollector) UpdateStats(stats *info.ContainerStats) error {
 	klog.V(5).Info("Attempting to update uncore perf_event stats")
 
-	for name, pmus := range c.cpuFiles {
-		for pmu, cpus := range pmus {
-			for cpu, file := range cpus {
-				stat, err := readPerfUncoreStat(file, name, cpu, pmu, c.topology)
+	for _, groupPMUs := range c.cpuFiles {
+		for pmu, group := range groupPMUs {
+			for cpu, file := range group.cpuFiles[group.leaderName] {
+				stat, err := readPerfUncoreStat(file, group, cpu, pmu, c.topology)
 				if err != nil {
-					return fmt.Errorf("unable to read from uncore perf_event_file (event: %q, CPU: %d, PMU: %s): %q", name, cpu, pmu, err.Error())
+					klog.Warningf("Unable to read from perf_event_file (event: %q, CPU: %d) for %q: %q", group.leaderName, cpu, pmu, err.Error())
+					continue
 				}
-				klog.V(5).Infof("Read uncore perf event (event: %q, CPU: %d, PMU: %s): %d", name, cpu, pmu, stat.Value)
 
-				stats.PerfUncoreStats = append(stats.PerfUncoreStats, *stat)
+				stats.PerfUncoreStats = append(stats.PerfUncoreStats, stat...)
 			}
 		}
 	}
 
 	return nil
 }
 
-func (c *uncoreCollector) setupRawNonGroupedUncore(event *CustomEvent, pmus []pmu) error {
-	klog.V(5).Infof("Setting up non-grouped raw perf uncore event %#v", event)
-
-	if event.Type == 0 {
-		// PMU isn't set. Register event for all PMUs.
-		for _, pmu := range pmus {
-			newEvent := CustomEvent{
-				Type:   pmu.typeOf,
-				Config: event.Config,
-				Name:   event.Name,
-			}
-			config := createPerfEventAttr(newEvent)
-			err := c.registerUncoreEvent(config, string(newEvent.Name), pmu.cpus, pmu.name)
-			if err != nil {
-				return err
-			}
-		}
-		return nil
-	} else {
-		// Register event for the PMU.
-		config := createPerfEventAttr(*event)
-		pmu, err := getPMU(pmus, event.Type)
-		if err != nil {
-			return err
-		}
-		return c.registerUncoreEvent(config, string(event.Name), pmu.cpus, pmu.name)
+func (c *uncoreCollector) setupEvent(name string, pmus uncorePMUs, groupIndex int, leaderFileDescriptors map[string]map[uint32]int) error {
```

Comment:
Shrink number of parameters. You may not need some of them if you decide to move `setGroupAttributes` call into `registerEvent`. I'm open to discussing it but I do believe that logic of selecting leader is opaque and it must be straightforward and kept in *one* place.